docs/design_proposals/sync-auto-support.md

```diff
@@ -0,0 +1,206 @@
+# Title
+
+* Author(s): Appu Goundan (@loosebazooka)
+* Design Shepherd: Tejal Desai (@tejal29), Priya Wadhwa (@priyawadhwa)
+* Date: 09/17/2019
+* Status: New
+
+## Background
+
+Currently skaffold does not support `sync` for files that are generated
+during a build. For example when syncing java files to a container, one
+would normally expect `.class` files to be sync'd, but skaffold is
+really only aware of `.java` files in a build.
+
+1. Why is this required?
+  - We would like to support the following features
+      - Skaffold can sync files that are not watched (these files are not
+        considered inputs for the container builder)
+      - Skaffold can sync files that are generated by a buildscript
+      - Skaffold can be configured to run a custom script to when certain files change
+
+2. If this is a redesign, what are the drawbacks of the current implementation?
+  - This is not a redesign, but a new sync mode that may or may not be available
+    in the skaffold.yaml directly to users. It could simply be an internal API
+    that is usuable by builders like Jib.
+  - This is similar to `_smart_` described in [sync-improvements](sync-improvements.md)
+
+3. Is there any another workaround, and if so, what are its drawbacks?
+  - Currently one can create a docker file that only copies specific build
+    results into the container and relies on a local build to generate those
+    intermediate artifacts. This requires a user to trigger a first build
+    manually before docker kicks in to containerize the application. While
+    it may be possible to automate this (for example: gradle --continuous), it
+    is not an acceptable solution to require manual external processes for
+    a build to succeed. Covered in [Hack it](#hack-it)
+
+4. Mention related issues, if there are any.
+  - This is not trying to solve the problem of dealing with a multistage
+    dockerbuild. Intermediate build artifacts might still be possible to
+    determine, however that would require an extra mechanism to do a partial
+    docker build and sync files from a built container to a running container --
+    something we do not intend to cover here.
+
+#### Problems with current API/config
+The current `sync` system has the following problems:
+1. No way to trigger local external processes - for example, in jib, to build a
+   container, one would run `./gradlew jib`, but to update class files so the
+   system may sync them, one would only be required to run `./gradlew classes`.
```

Comment:
Does Cloud Code (or IntelliJ in general) puts build outputs into the usual `build/` or `target/` directory? (Eclipse doesn't; it outputs to a separate directory that doesn't interfere with command-line Gradle/Maven.) Also, does IntelliJ actually runs command-line Gradle/Maven by default when the user modifies a Java source file inside the IDE? (I'm guessing no for the latter question.) I'm asking because IDEs will compile `.class` files on-the-fly when the source files are modified inside the IDE and places into the default build output directory, the user won't actually need to run `./gradlew classes`.

Reply:
Yeah, so I don't think we'll be able to solve the ide/build-system cooperation problem here. What we expect is that saved source changes will trigger a rebuild+sync through skaffold-jib. What the ide does might be independent/irrelevant.

example: 
- user changes MyClass.java
  - ide compiles and outputs to its own output directory
    - nothing further happens related to the IDE
  - skaffold sees `MyClass.java` has changed
    - skaffold triggers a `gradle classes` to build classes
    - skaffold syncs classes with remote container.